dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -322,37 +328,102 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
 
-        unsigned int count{};
-        if (toastPropertiesCollection)
+                toastNotifications.Append(toastNotification);
+            }
+
+            co_return toastNotifications;
+        }
+        catch (...)
         {
-            THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            eptr = std::current_exception();
+            telemetryResult = wil::ResultFromCaughtException();
         }
+    }
+#else // Option 2
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification>> AppNotificationManager::GetAllAsync()
+    {
+        std::wstring appId{};
 
-        winrt::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification> toastNotifications{ winrt::single_threaded_vector<winrt::Microsoft::Windows::AppNotifications::AppNotification>() };
-
-        for (unsigned i = 0; i < count; ++i)
+        try
         {
-            ToastABI::INotificationProperties* toastProperties;
-            THROW_IF_FAILED(toastPropertiesCollection->GetAt(i, &toastProperties));
+            co_await winrt::resume_background();
 
-            auto toastNotification{ ToastNotificationFromToastProperties(toastProperties) };
+            appId = RetrieveNotificationAppId();
 
-            toastNotifications.Append(toastNotification);
-        }
+            winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection{};
+            auto result{ ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()) };
+
+            THROW_HR_IF(result, result != S_OK && result != E_NOT_SET); // Swallow E_NOT_SET and return an empty properties vector to signal that there are no active toasts
+
+            unsigned int count{};
+            if (toastPropertiesCollection)
+            {
+                THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            }
+
+            winrt::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification> toastNotifications{ winrt::single_threaded_vector<winrt::Microsoft::Windows::AppNotifications::AppNotification>() };
+
+            for (unsigned i = 0; i < count; ++i)
+            {
+                ToastABI::INotificationProperties* toastProperties;
+                THROW_IF_FAILED(toastPropertiesCollection->GetAt(i, &toastProperties));
+
+                auto toastNotification{ ToastNotificationFromToastProperties(toastProperties) };
 
-        co_return toastNotifications;
+                toastNotifications.Append(toastNotification);
+            }
+
+            AppNotificationTelemetry::GetAllAsyncByAPI(S_OK, appId);
+
+            co_return toastNotifications;
+        }
+        catch (...)
+        {
+            AppNotificationTelemetry::GetAllAsyncByAPI(wil::ResultFromCaughtException(), appId);
+            throw;
```

Comment:
It doesn't seem like this `throw;` propagate the error code? Maybe we can use `THROW_HR(wil::Result...())`.

Reply:
Throw propagates the current caught exception. The error code is part of the exception, there is no point in extracting the error code from the exception and creating a new exception based on that same error code.

Reply:
Think about it this way: we only catch the exception because we want to get some telemetry, otherwise we wouldn't catch it. That's exactly what the code does: catch the exception, extract and log the error code, then let the exception propagate without additional intervention.

Reply:
I see. Thanks for explaining!